Extract MM_GlobalCollectorDelegate from CLI

ktbriggs opened the pull request:
This is ready for review. Flagging WIP (complete) pending review comments.
 
Extract from MM_CollectorLanguagteInterface (CLI) a delegate class 
MM_GlobalCollectorDelegate encapsulating the CLI methods:

doFrequentObjectAllocationSampling(..)
checkForExcessiveGC(..)
globalCollector_isTimeForGlobalGCKickoff(..)
globalCollector_internalPostCollect(..)
parallelGlobalGC_masterThreadGarbageCollect_beforeGC(..)
parallelGlobalGC_masterThreadGarbageCollect_afterGC(..)
parallelGlobalGC_postPrepareHeapForWalk(..)
parallelGlobalGC_postMarkProcessing(..)
parallelGlobalGC_setupBeforeGC(..)
parallelGlobalGC_setMarkingScheme(..)
parallelGlobalGC_createAccessBarrier(..)
parallelGlobalGC_destroyAccessBarrier(..)
parallelGlobalGC_heapAddRange(..)
parallelGlobalGC_heapRemoveRange(..)
parallelGlobalGC_createHeapWalker(..)
parallelGlobalGC_destroyHeapWalker(..)
parallelGlobalGC_getHeapWalker(..)
parallelGlobalGC_checkIfCompactionShouldBePrevented(..)
parallelGlobalGC_masterThreadGarbageCollect_gcComplete(..)
parallelGlobalGC_collectorInitialized(..)
parallelGlobalGC_reportObjectEvents(..)
heapWalker_heapWalkerObjectSlotDo(..)
parallelDispatcher_handleMasterThread(..)

These methods have been removed from the CLI and moved to 
MM_GlobalCollectoDelegate or merged into existing OMR code.

See example/glue/GlobalCollectorDelegate.hpp.

OMR Issue #1048
Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

ktbriggs commented:
Ready for review.

charliegracie commented on gc/base/Collector.cpp:
```diff
@@ -237,6 +260,155 @@ MM_Collector::preCollect(MM_EnvironmentBase* env, MM_MemorySubSpace* subSpace, M
+
+	OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
+	TRIGGER_J9HOOK_MM_PRIVATE_EXCESSIVEGC_CHECK_GC_ACTIVITY(extensions->privateHookInterface,
+															env->getOmrVMThread(),
```
Can you reduce the tabbing here to make this more read-able?  Thanks.

charliegracie commented on gc/base/Collector.cpp:
```diff
@@ -237,6 +260,155 @@ MM_Collector::preCollect(MM_EnvironmentBase* env, MM_MemorySubSpace* subSpace, M
+			extensions->excessiveGCLevel = excessive_gc_fatal;
+
+			TRIGGER_J9HOOK_MM_OMR_EXCESSIVEGC_RAISED(extensions->omrHookInterface,
+												 env->getOmrVMThread(),
```
Can you reduce the tabbing here to make this more read-able?  Thanks.

charliegracie commented on gc/base/Collector.cpp:
```diff
@@ -237,6 +260,155 @@ MM_Collector::preCollect(MM_EnvironmentBase* env, MM_MemorySubSpace* subSpace, M
+		reclaimedPercent = ((float)heapFreeDelta / (float)extensions->heap->getActiveMemorySize()) * 100;
+
+		TRIGGER_J9HOOK_MM_PRIVATE_EXCESSIVEGC_CHECK_FREE_SPACE(extensions->privateHookInterface,
+															   env->getOmrVMThread(),
```
Can you reduce the tabbing here to make this more read-able?  Thanks.

charliegracie commented on gc/base/Collector.cpp:
```diff
@@ -237,6 +260,155 @@ MM_Collector::preCollect(MM_EnvironmentBase* env, MM_MemorySubSpace* subSpace, M
+			Trc_MM_ExcessiveGCRaised(env->getLanguageVMThread());
+
+			TRIGGER_J9HOOK_MM_OMR_EXCESSIVEGC_RAISED(extensions->omrHookInterface,
+												 env->getOmrVMThread(),
```
Can you reduce the tabbing here to make this more read-able?  Thanks.

ktbriggs replied:
Done (tabs in Collector.cpp).

charliegracie commented on gc/base/GlobalCollector.hpp:
```diff
@@ -36,9 +37,23 @@ class MM_MemorySubSpace;
+	 * Informational method called when master thread is started, to allow language-specific action is required. Calls
+	 * MM_GlobalCollectorDelegate::startMasterThread() to effect language specialization.
+	 */
+	virtual void startMasterThread(MM_EnvironmentBase *env) { _delegate.startMasterThread(env); }
```
This is only used for Global collectors?  Generational and Incremental collectors do not need this functionality? This feels like something that may belong in a different class and not the collector hierarchy.

ktbriggs replied:
Thanks for catching this. This java artifact relates to ParallelDispatcher, which is used in contexts other than standard global collection. Will factor this into MM_EnvironmentBase::setThreadType() to allow use in these contexts. Also, will tweak ParallelDispatcher to ensure that MM_EnvironmentBase::setThreadType() is informed when master thread completes GC (relinquishing master thread status).

charliegracie commented on gc/base/GlobalCollector.hpp:
```diff
@@ -84,6 +99,7 @@ class MM_GlobalCollector : public MM_Collector {
 
 	MM_GlobalCollector(MM_EnvironmentBase* env, MM_CollectorLanguageInterface *cli)
 		: MM_Collector(cli)
+	, _delegate()
```
Can you tab this to align it with the other fields? Thanks

ktbriggs replied:
Done.

charliegracie commented on gc/base/ParallelHeapWalker.cpp:
```diff
@@ -0,0 +1,162 @@
+/*******************************************************************************
```
For all of the new files in the GC folder directly are they newly written code or are they being donated by the J9 project?

ktbriggs replied:
GlobalCollectorDelegate.*pp are new files that contain previously written (refactored) code, much of which was originally contributed from J9. The following files were migrated from (contributed by) J9:

MarkMapSegmentChunkIterator.*pp
ParallelHeapWalker.*pp
ParallelObjectheapIterator.*pp
ParallelGlobalGCTask.*pp

charliegracie commented on gc/base/standard/HeapWalker.cpp:
```diff
@@ -93,8 +94,10 @@ heapWalkerObjectSlotDo(OMR_VMThread *omrVMThread, MM_HeapRegionDescriptor *regio
 	 * OK. We should simplify this code once classes-on-heap is enabled
```
Can you clean up the comment while you are making changes to remove the references to classes and talk about other off heap memory or fields that need to be updated?  What if there is more than one indirect reference from the object?

ktbriggs replied:
Done (comment cleanup). Good point about indirect objects. Scavenger needs to deal with this as well, and also allows for only one indirect (unreachable from reference graph) object per reachable heap object. I opted for this to avoid requiring a second type of object scanner/iterator (eg, getIndirectObjectIterator(omrobjectptr_t). Perhaps we need one but current model supports Java/example and I've not heard of any new requirements from other languages/frameworks.

ktbriggs commented:
b6c81c1 addresses review comments and rebases 3d0c5b9 onto eclipse/master.

ktbriggs commented:
43952c0 rebases previously reviewed commit onto eclipse/master; 3627925 addresses review comments, and pulls into ParallelGlobalGC.cpp some changes relating to concurrent scavenger (conflicts from rebasing).

ktbriggs commented:
@charliegracie let me know when you have finished reviewing 3627925 changes and I will squash/push for merge.

charliegracie commented:
looks good to me so you can squash.  Before you push the squashed change though can you separate the changes into 2 commits?  One commit for contributing the files from J9 to OMR and another for the rest of the changes.  Thanks.

ktbriggs commented:
Ok. 3519c5d squashes previous 2 commits , excluding files migrated from J9. J9 files incorporated into OMR are in c9e2a14.
